Fix `DoubleProposalProof`s

hrxi opened the pull request:
They required data unrelated to the signed proposals. It now uses the same signing infrastructure by factoring it out to `TendermintProposal` in a common crate.

Fixes #2981.

hrxi committed:
Fix `DoubleProposalProof`s

They required data unrelated to the signed proposals. It now uses the
same signing infrastructure by factoring it out to `TendermintProposal`
in a common crate.

Fixes #2981.

hrxi commented on primitives/block/src/equivocation_proof.rs:
```diff
@@ -405,13 +419,27 @@ impl DoubleProposalProof {
 
         if self.header1.block_number != self.header2.block_number
             || self.header1.round != self.header2.round
```
Is this check still correct or should it go away?

nibhar replied:
I would say this was not correct before and it should be removed. The round in the header only denotes when a proposal was created, not when it was proposed. Thus it should not be checked here.

hrxi commented on primitives/block/src/equivocation_proof.rs:
```diff
@@ -327,20 +331,30 @@ impl DoubleProposalProof {
         mut justification1: SchnorrSignature,
         mut header2: MacroHeader,
+        mut round2: u32,
+        mut valid_round2: Option<u32>,
```
Should I merge the two `round` parameters into one, given that they have to be the same?

nibhar replied:
I would prefer these be `TendermintProposal`, making the signature much smaller. If that turns out to be annoying or to require unnecessary cloning, then I think one round could be removed yes.

nibhar commented on primitives/block/src/equivocation_proof.rs:
```diff
@@ -317,6 +317,10 @@ pub struct DoubleProposalProof {
     header1: MacroHeader,
     /// Header number 2.
     header2: MacroHeader,
+    round1: u32,
+    round2: u32,
+    valid_round1: Option<u32>,
+    valid_round2: Option<u32>,
```
I would prefer
```suggestion
    proposal1: TendermintProposal,
    /// Proposal number 2.
    proposal2: TendermintProposal,
```

nibhar reviewed (changes requested):
Looks good to me outside of the `MacroHeader.round` check being incorrect (existed prior to this PR). So that should be removed. Other comments are purely cosmetic .

hrxi committed:
Allow `#[derive(SerializedMaxSize, SerializedSize)]` on generic types

hrxi committed:
Use `TendermintProposal` in `DoubleProposalProof`

hrxi committed:
Don't check round in block header of double proposal proofs

The round in the proposal message is the relevant one.

nibhar commented on primitives/block/src/equivocation_proof.rs:
```diff
@@ -326,21 +326,21 @@ pub struct DoubleProposalProof {
-        let hash1: Blake2bHash = header1.hash();
-        let hash2: Blake2bHash = header2.hash();
+        let hash1: Blake2bHash = proposal1.proposal.hash();
+        let hash2: Blake2bHash = proposal2.proposal.hash();
```
Maybe we should add an assert for the equal rounds here. Not strictly necessary but `DoubleProposalProof`s without matching rounds are invalid and should never be created.
```suggestion
        assert_eq!(proposal1.round, proposal2.round);
        let hash1: Blake2bHash = proposal1.proposal.hash();
        let hash2: Blake2bHash = proposal2.proposal.hash();
```

hrxi replied:
Not done for now because we currently don't do that for the creation of any equivocation proof structure. We could automatically call `verify` I guess and `assert!` that it verifies. But maybe we could also do that at the call sites?

nibhar commented on primitives/block/src/equivocation_proof.rs:
```diff
@@ -360,36 +360,36 @@ impl DoubleProposalProof {
     ) -> Result<(), EquivocationProofError> {
-        let hash1: Blake2bHash = self.header1.hash();
-        let hash2: Blake2bHash = self.header2.hash();
+        let hash1: Blake2bHash = self.proposal1.proposal.hash();
+        let hash2: Blake2bHash = self.proposal2.proposal.hash();
```
As far as I can see these are now only used for sorting. Sorting could also just use the Blake2sHash instead saving one hashing.

hrxi replied:
Done.

nibhar reviewed (approved):
LGTM. Left minor comments only.

hrxi committed:
Use `TendermintProposal` for internal sorting of `DoubleProposalProof`

This means that two different `TendermintProposal`s with the same
`MacroHeader` are now also considered a violation.

hrxi commented:
Rebased for merging.
